File: 621-task-scheduler/621-task-scheduler.scala
import scala.collection.mutable._
object Solution {
    def leastInterval(tasks: Array[Char], n: Int): Int = {
        val bucket: Map[Char, Int] = Map()
        var m = 0
        var s = 0

        tasks.foreach { task =>
            val num = bucket.get(task).getOrElse(0) + 1
            
            bucket += (task -> num)
            if (num == m)  s += 1
            else if (num > m) { 
                m = num
                s = 1
            }
        }

        ((n + 1) * (m - 1) + 1) + (s - 1) max tasks.length
    }
}

File: 621-task-scheduler/621-task-scheduler.java
class Solution {
    public int leastInterval(char[] tasks, int n) {
        int[] occurence = new int[26];
        for(char t : tasks) occurence[t - 'A']++;
        Arrays.sort(occurence);
        
        int idle = (occurence[25] - 1) * n;
        
        int i = 24;
        while(idle > 0 && i >= 0) {
            idle -= Math.min(occurence[25] - 1, occurence[i--]);
        }
        
        idle = Math.max(0, idle);
        
        return tasks.length + idle;
    }
}